packages/material-ui-lab/src/TreeView/TreeView.js

```diff
@@ -203,16 +203,21 @@ const TreeView = React.forwardRef(function TreeView(props, ref) {
     setExpandedState(newExpanded);
   };
 
+  const isExpandable = id => {
```

Comment:
We don't need this logic. We can apply the same logic that is used in `aria-expanded`:

```diff
diff --git a/packages/material-ui-lab/src/TreeItem/TreeItem.js b/packages/material-ui-lab/src/TreeItem/TreeItem.js
index 1b3c568b7..f423141e7 100644
--- a/packages/material-ui-lab/src/TreeItem/TreeItem.js
+++ b/packages/material-ui-lab/src/TreeItem/TreeItem.js
@@ -368,7 +368,7 @@ const TreeItem = React.forwardRef(function TreeItem(props, ref) {
   return (
     <li
       className={clsx(classes.root, className, {
-        [classes.expanded]: expanded,
+        [classes.expanded]: expandable && expanded,
         [classes.selected]: selected,
       })}
       role="treeitem"
```

Reply:
Disagree.

Although your change applies the correct class shouldn't the expanded state be correct in its own right ?

Additionally:

In the case of expandAllSiblings, without the change, the onNodeToggle event will have end nodes in the expanded argument.

handlePreviousArrow is relying on the correct value for expandable
```
const handlePreviousArrow = event => {
    if (expanded) {
      toggleExpansion(event, nodeId);
      return true;
    }

    const parent = getParent(nodeId);
    if (parent) {
      focus(parent);
      return true;
    }
    return false;
  };
```

Reply:
There is also an issue with a dynamic tree

```
React.useEffect(() => {
    const buildVisible = nodes => {
      let list = [];
      for (let i = 0; i < nodes.length; i += 1) {
        const item = nodes[i];
        list.push(item);
        const childs = nodeMap.current[item].children;
        if (isExpanded(item) && childs) {
          list = list.concat(buildVisible(childs));
        }
      }
      return list;
    };

    if (childrenCalculated) {
      visibleNodes.current = buildVisible(nodeMap.current[-1].children);
    }
  }, [expanded, childrenCalculated, isExpanded]);
```

When all children are removed from a TreeItem the code above does not execute ( as expanded is not updated ) and visibleNodes contains nodes that have been removed.  
This [pull](https://github.com/mui-org/material-ui/pull/20157) would resolve this issue.

You can see this here https://codesandbox.io/s/inspiring-platform-xrbk5
1. Focus cabbage
2. Click 'Make End Node'
3. Tab to cabbage
4. Arrow Up
5. Arrow Up - cannot get to fruits

You can also see another major issue by doing the following

1. Focus apples
2. Click make end node - **now none of the tree items are tabbable** 

We should be performing more cleanup when removeNodeFromNodeMap as I do in this [pull](https://github.com/mui-org/material-ui/pull/20085/files) 
https://github.com/mui-org/material-ui/blob/98835c6082b7afe4c1258a4b8968afd3ee591221/packages/material-ui-lab/src/TreeView/TreeView.js#L426

I will not create an issue for this focus issue as I will be creating a pull for https://github.com/mui-org/material-ui/issues/20097 as I think our focus behaviour is not entirely correct currently.

> When a single-select tree receives focus:
If none of the nodes are selected before the tree receives focus, focus is set on the first node.
If a node is selected before the tree receives focus, focus is set on the selected node.
When a multi-select tree receives focus:
If none of the nodes are selected before the tree receives focus, focus is set on the first node.
If one or more nodes are selected before the tree receives focus, focus is set on the first selected node.